src/core/Instrument.cpp

```diff
@@ -78,8 +82,96 @@ bool Instrument::isFromTrack( const Track * _track ) const
 	return( m_instrumentTrack == _track );
 }
 
+// helper function for Instrument::applyFadeIn
+static int countZeroCrossings(sampleFrame *buf, fpp_t start, fpp_t frames)
+{
+	// zero point crossing counts of all channels
+	int zeroCrossings[DEFAULT_CHANNELS] = {0};
+	// maximum zero point crossing of all channels
+	int maxZeroCrossings = 0;
+
+	// determine the zero point crossing counts
+	for (fpp_t f = start; f < frames; ++f)
+	{
+		for (ch_cnt_t ch=0; ch < DEFAULT_CHANNELS; ++ch)
+		{
+			// we don't want to count [-1, 0, 1] as two crossings
+			if ((buf[f-1][ch] <= 0.0 && buf[f][ch] > 0.0) ||
+					(buf[f-1][ch] >= 0.0 && buf[f][ch] < 0.0))
+			{
+				++zeroCrossings[ch];
+				if (zeroCrossings[ch] > maxZeroCrossings)
+				{
+					maxZeroCrossings = zeroCrossings[ch];
+				}
+			}
+		}
+	}
+
+	return maxZeroCrossings;
+}
+
+// helper function for Instrument::applyFadeIn
+fpp_t getFadeInLength(float maxLength, fpp_t frames, int zeroCrossings)
+{
+	// calculate the length of the fade in
+	// Length is inversely proportional to the max of zeroCrossings,
+	// because for low frequencies, we need a longer fade in to
+	// prevent clicking.
+	return (fpp_t) (maxLength  / ((float)zeroCrossings / ((float)frames/128.0f) + 1.0f));
+}
+
 
+void Instrument::applyFadeIn(sampleFrame * buf, NotePlayHandle * n)
+{
+	const static float MAX_FADE_IN_LENGTH = 85.0;
```

Comment:
Do we want this constant here or in a header?

Reply:
To further elaborate, do we want this to be customizable per `Instrument`?

Reply:
> Do we want this constant here or in a header?

It's matter of preference. I think it's better to put this here, because this constant is used only in this method.

> To further elaborate, do we want this to be customizable per `Instrument`?

Good idea. But we need to consider the possibility that end users might find it too technical. Most users don't want clicks, however it's better to give options to advanced users.